### PR TITLE
fix: upgrade to newer react-pdf and pdf.js to render high-resolution images properly on the web

### PR DESCRIPTION
**Pull request by artus9033**
### Details

PDF files carrying images of large resolutions (such as the one posted in this issue) are not rendered on iOS Safari, although we would like them to be rendered properly.
This is because our app on the web uses `react-pdf`, which in turn uses `pdf.js` (here, from an NPM package `pdfjs-dist`) as renderer. The renderer utilizes a canvas to render PDF contents. Every browser has limitations applicable to the canvas element, since it requires a buffer to be allocated in memory. [It is specified](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/canvas#maximum_canvas_size) and [well tested](https://github.com/jhildenbiddle/canvas-size#test-results) that Safari limits the maximum size to 4096x4096 px. The [PDF provided in the original issue](https://github.com/Expensify/App/files/13567151/Bug6302114_1701797701028.pdf.pdf) carries an image that breaks this limit.

The maintainers of `pdf.js` commented that starting from version 3.5.141 [they are attempting](https://github.com/mozilla/pdf.js/issues/16680#issuecomment-1632504025) to use the `OffscreenCanvas` to downsize over-large images in PDFs, if it is implemented in the host browser at runtime, which fixes the issue at the expense of rendering quality.

This PR concentrates on porting `src/components/PDFView` to work with `react-pdf` version `7.3.3` (which relies on `pdfjs-dist` version `3.6.172` that contains the aforementioned feature) that's aligned with the current work done in (WIP) switch to a custom library [here](https://github.com/Expensify/App/issues/22998).

### Fixed Issues

$ https://github.com/Expensify/App/issues/32502
PROPOSAL: https://github.com/Expensify/App/issues/32502#issuecomment-1849595362

### Tests

- Open a chat
- Send the [PDF file](https://github.com/Expensify/App/files/13567151/Bug6302114_1701797701028.pdf.pdf) attached in the original issue, containing a high resolution image
- Open the PDF
- Ensure that after a while (rendering) the image appears, in a slightly lower quality due to downscaling
- Ensure that other PDFs still work after changes, especially [this large document](https://github.com/Expensify/App/files/11639141/LARGE.PDF.FILE.pdf) from [this issue](https://github.com/Expensify/App/issues/19918) that makes use of virtualized list rendering
- Ensure that PDFs secured with passwords are handled properly
- Ensure that there are no redundant re-renders of the PDF view (i.e., the PDF does not 'flash' when performing an interaction such as clicking / scrolling)

### Offline tests

N/A since loading of chat attachments requires an internet connection.

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/10586109/7b53f423-d087-4249-99e8-1e7fc58a4d4a

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/10586109/885fac63-2f08-4dca-ac97-5ee9285d2c11

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/10586109/a7da31ea-94ce-4100-a359-ff8e5d21cd11

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/10586109/5a541169-4444-4d8f-bfd7-1b3b2170226f

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/10586109/5749d24c-7c30-4b09-aaf4-55533be945b8

https://github.com/Expensify/App/assets/10586109/92447982-1d0e-4858-a8a9-af725532c1a2

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/10586109/0788f5eb-0239-4afe-87f2-7461b401fe47
</details>

### CODE REVIEW

File: src/components/PDFView/ReactPDFPageRenderer.js
Comment (rezkiy37) [inline src/components/PDFView/ReactPDFPageRenderer.js:8]: Just rename.
```suggestion
const PDFPageRenderer = memo(({index, data, style}) => {
```

Comment (artus9033) [reply src/components/PDFView/ReactPDFPageRenderer.js:8]: Thanks, indeed the destructuring with alias is a great idea. Regarding the naming convention, do you have in mind another candidate? I wanted to somehow mark that this is a web-only component, since for native we have a different library.

Comment (rezkiy37) [reply src/components/PDFView/ReactPDFPageRenderer.js:8]: I got your idea. Let's use the clearest name `WebPDFPageRenderer` 🙂

Comment (artus9033) [reply src/components/PDFView/ReactPDFPageRenderer.js:8]: Sounds good to me 😄

Comment (artus9033) [reply src/components/PDFView/ReactPDFPageRenderer.js]: Resolved in https://github.com/Expensify/App/pull/33083/commits/d9f90c12789e410017e04a75dae7db45267649af

**Comment by jjcoffee** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by jjcoffee** [comment]
@artus9033 Thanks for the PR! I've just come across a few of issues whilst testing.

1. Navigating back and forth causes an ugly flash/reload on the large-image pdf:

https://github.com/Expensify/App/assets/27287420/cb114b8d-56ab-42b8-b37f-a4ce16c7eae7

2. Password-protected PDFs don't seem to load:

https://github.com/Expensify/App/assets/27287420/b66c95e0-5e32-408c-a5b7-1bc272bd8055

3. Console error when resizing the browser window:

image

**Comment by jjcoffee** [comment]
Similar to the first issue, but on mobile it's more obvious as the image preview disappears as you swipe or if you just tap the screen. Also getting another crash (I don't have 100% reliable repro steps but it seems to happen after tapping a few times and then swiping back and forth), console error below.

https://github.com/Expensify/App/assets/27287420/dc98ecd5-2624-439e-8051-0a961021e537

image

**Comment by artus9033** [comment]
Thank you @jjcoffee, sorry for the trouble - I will track & fix the issues you mentioned.

File: src/components/PDFView/WebPDFDocument.js
Comment (rezkiy37) [inline src/components/PDFView/WebPDFDocument.js:104]: Create a `defaultProps` variable before the component.

File: src/components/PDFView/WebPDFDocument.js
Comment (rezkiy37) [inline src/components/PDFView/WebPDFDocument.js:101]: Let's apply `undefined` by default.

File: src/components/PDFView/index.js
Comment (rezkiy37) [inline src/components/PDFView/index.js:243]: Let's simplify this. We can assign `[styles.PDFView, styles.noSelect, this.props.style]` by default and `push` additional style `styles.invisible` if it is required.

File: src/components/PDFView/WebPDFDocument.js
Comment (artus9033) [reply src/components/PDFView/WebPDFDocument.js]: Thanks, resolved in 44fc03995053004beca12a70523fe45abe9e0653

File: src/components/PDFView/WebPDFDocument.js
Comment (artus9033) [reply src/components/PDFView/WebPDFDocument.js]: Thanks, resolved in 44fc03995053004beca12a70523fe45abe9e0653

File: src/components/PDFView/WebPDFDocument.js
Comment (artus9033) [reply src/components/PDFView/WebPDFDocument.js]: Right, that doesn't change any behaviour and is still consistent - resolved in 44fc03995053004beca12a70523fe45abe9e0653

File: src/components/PDFView/index.js
Comment (artus9033) [reply src/components/PDFView/index.js]: Totally agreed, resolved in 67ab4a531d65c6b777de1e6b54050101bc0d16d2

**Comment by artus9033** [comment]
@jjcoffee the updated PR is ready for review now  🚀

**Comment by jjcoffee** [comment]
@artus9033 Thanks and sorry for the delay (I was OOO). Could you resolve conflicts?

**Comment by jjcoffee** [comment]
@artus9033 Is there a way to reduce rerenders on window resize?

**Comment by jjcoffee** [comment]
@artus9033 There's a brief flash of a zero on iOS safari before the large image PDF loads:

image

**Comment by jjcoffee** [comment]
@artus9033 Friendly bump on the above!

**Comment by MrRefactor** [comment]
> @artus9033 Friendly bump on the above!


Hi @jjcoffee Im taking this PR from Artur and will resolve the issue you mentioned soon!

**Comment by artus9033** [comment]
Right, @MrRefactor will handle it from here 🔧

**Comment by MrRefactor** [comment]
@dylanexpensify I should be able to add the fix tomorrow

**Comment by MrRefactor** [comment]
Sorry, I wasnt able to post to fix yet, will try to do it as soon as possible

**Comment by MrRefactor** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by MrRefactor** [comment]
recheck

**Comment by MrRefactor** [comment]
@jjcoffee Hi, I have posted the fix to the PR, but at the moment Im struggling with running the app after merging changes from main - for both this branch and main. I hope to fix that issue with my setup tomorrow morning so I could resolve conflicts and push it.

**Comment by jjcoffee** [comment]
@MrRefactor Hey, thanks for the update! Please can you also make sure to sign all your commits per [the guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#begin-coding-your-solution-in-a-pull-request)?

**Comment by MrRefactor** [comment]
Sure thing, fixing right now @jjcoffee

**Comment by MrRefactor** [comment]
recheck

**Comment by MrRefactor** [comment]
@jjcoffee I have just signed all the commits, the question is - where should I put updated videos? I also believe that checklist changed after Artur created this PR so how should we adjust that?

**Comment by jjcoffee** [comment]
Any way to stop the screen going white when resizing the window? I see it also happens on staging so we don't necessarily need to handle it here, but it would be nice if it's straight-forward enough:

https://github.com/Expensify/App/assets/27287420/5c82c50a-c13c-45c6-a166-3425bf7988a6

**Comment by MrRefactor** [comment]
> Any way to stop the screen going white when resizing the window? I see it also happens on staging so we don't necessarily need to handle it here, but it would be nice if it's straight-forward enough:
> 
>  chrome-resize-2024-01-17_14.29.55.mp4

I dont think we are able to fix it right now - this behaviour is strictly connected to the flow of rendering pdf itself while resizing view -> resizing canvas size -> calculating density of pixels -> and then rerendering

**Review by jjcoffee — APPROVED** [review]
Tests well & LGTM!

**Comment by jjcoffee** [comment]
Checklist updated to the latest version.

**Comment by danieldoglas** [comment]
@jjcoffee shouldn't the new files be created in TS already?

**Comment by jjcoffee** [comment]
@danieldoglas Hmm good point! I think since the PR was raised before the announcement it might be best to avoid extra complexity by migrating the existing PR to TS as well. @MrRefactor just added a small tweak to @artus9033's original PR.

Once we switch to our [custom PDF library](https://github.com/Expensify/App/issues/22998), I'd expect all of that to be written in TS, maybe @rezkiy37 can confirm?

**Comment by rezkiy37** [comment]
> I'd expect all of that to be written in TS, maybe @rezkiy37 can confirm?

The [library](https://github.com/Expensify/react-fast-pdf) is being written in **TS**. So, I already started to integrate `react-fast-pdf` into **Expensify**: https://github.com/Expensify/App/pull/33434, https://github.com/Expensify/react-fast-pdf/pull/12. Currently, I am waiting on this PR to be merged to continue working.

cc @jjcoffee

**Comment by jjcoffee** [comment]
@danieldoglas Sounds like we're good to merge this so that we unblock @rezkiy37's work on `react-fast-pdf`.

**Comment by danieldoglas** [comment]
@jjcoffee can you please use the most updated version of the checklist, since this is not passing the test for it?

Also, I'm trying to understand why this is not passing the snyk tests

**Comment by jjcoffee** [comment]
@danieldoglas Already updated, the check just needs to be rerun, I think!

**Review by danieldoglas — DISMISSED** [review]

**Comment by danieldoglas** [comment]
@rezkiy37 the snyk error is the following:

```
Out of sync package.json and package-lock.json detected. Please run "package.json install" and try again.
```


Can you please fix it and push it?

**Comment by rezkiy37** [comment]
> @rezkiy37 the snyk error is the following:
> 
> ```
> Out of sync package.json and package-lock.json detected. Please run "package.json install" and try again.
> ```
> 
> Can you please fix it and push it?

Sorry, but current PR it is not mine. 
cc @MrRefactor

**Comment by jjcoffee** [comment]
@danieldoglas Apparently the snyk error happens on all PRs where the package file is changed, see [here](https://github.com/Expensify/App/pull/32471#issuecomment-1845870033). So I think we're good to ignore for now!

**Review by danieldoglas — APPROVED** [review]

**Pull request merged**
